src/core/cgroup.c

```diff
@@ -3805,14 +3805,18 @@ int unit_get_memory_available(Unit *u, uint64_t *ret) {
                 if (!slice_context)
                         continue;
 
-                if (slice_context->memory_max == UINT64_MAX && slice_context->memory_high == UINT64_MAX)
+                if (unit_has_name(slice, SPECIAL_ROOT_SLICE))
+                        slice_limit = physical_memory();
```

Comment:
hmm, so I appreciate this change, but it's too simple in two cases:

1. systemd runs in a container, with or without cgroup namespacing. the root slice in that case is not managing the host, hence using physical memory limits doesn't really work

2. systemd runs in --user mode, similar case.

I figure this check here should hence be conditionalized:

```c
if (unit_has_name(slice, SPECIAL_ROOT_SLICE) && 
        u->manager->runtime_scope == RUNTIME_SCOPE_SYSTEM && 
        detect_container() == 0) {
```

Reply:
Even container or user instance are limited by `physical_memory()`. The `unit_get_memory_available` will take min of that (if available) or what's configured on the apparent root cgroup.
(Actually, `physical_memory` already takes the minimum of them too. So it's overly conservative here but it should still give appropriate result.)